### ISSUE TITLE
Project Level 5 

### ISSUE THREAD

**Issue by hengyjj** [nus-cs2030/2122-s2#247]
# Summary
Put a summary of the question

## Description 
Hello! I am currently at Level 5 of the Project. I would like to seek help on the issue whereby we are supposed to store the serviceTime within a lazy class as I am unsure on how to approach this. 

Within my Simulate5.java, I take in a number of servers and a `List<Pair<Double, Supplier<Double>>> arrivalServiceTimes.`

<img width="754" alt="image" src="https://user-images.githubusercontent.com/77519191/161727963-7a8c2122-dd53-4674-89ce-fe71a68c6721.png">

For now, within my `run()` method, there is a for loop where I would populate my PQ<Event>, and I will implement the `Lazy<Double> lazy = Lazy.<Double>of(this.arrivalServiceTimes.get(i).second()).`

<img width="645" alt="image" src="https://user-images.githubusercontent.com/77519191/161728529-6aed1209-4314-4dc1-8ec1-2bc1135c79d6.png">

However, this means that the code would always ask for the service time which isn't right as for customers who are leaving, there is no need to ask for the service time. However, for my code, the execution of the event is done below so I am not too sure how to go about this.
 
<img width="535" alt="image" src="https://user-images.githubusercontent.com/77519191/161728351-a16eb5fd-ed0d-44bf-86cb-b0d261ac108c.png">

Would appreciate any advices/help on this level. Thank you!

## Screenshots (if any): 
Insert Images here if necessary

**Comment by LauWyeRock** [comment]
you can consider making a new constructor for customer that takes in a supplier, and only when you are creating your serve or wait events you can call .get(). 

you shouldn't be calling .get() when you are populating your queue, just giving the customers the access to the supplier is sufficient. The assignment of service times would happen after you determine whether the customer leaves or not

**Comment by hengyjj** [comment]
> you can consider making a new constructor for customer that takes in a supplier, and only when you are creating your serve or wait events you can call .get().
> 
> you shouldn't be calling .get() when you are populating your queue, just giving the customers the access to the supplier is sufficient. The assignment of service times would happen after you determine whether the customer leaves or not

Hi WyeRock, thanks for your reply! I created a new constructor inside Customer that takes in `Supplier<Double> serviceTime`. However, I do not need to create any Lazy variables? 

My current code looks like this. 
<img width="766" alt="image" src="https://user-images.githubusercontent.com/77519191/161770759-d8a5e104-7afa-4ca4-b0cb-a0ae2a7e7599.png">

**Comment by LauWyeRock** [comment]
for level 5 your simulate5 should take in an int and a list of pair<Double, supplier<double>>.

to create my customers, i looped through the list and used the first double as arrival time, while the supplier is passed into a Lazy.<Double>of(supplier). Only when i'm deciding whether the customer leaves or not do I call .getServiceTime() which gets the service time using the supplier. (My customer has a Lazy<Double> serviceTime attribute)

**Comment by hengyjj** [comment]
> for level 5 your simulate5 should take in an int and a list of pair<Double, supplier>.
> 
> to create my customers, i looped through the list and used the first double as arrival time, while the supplier is passed into a Lazy.of(supplier). Only when i'm deciding whether the customer leaves or not do I call .getServiceTime() which gets the service time using the supplier. (My customer has a Lazy serviceTime attribute)

In this case, am I right to take in this instead? 
```
public Customer(int custID, double arrivalTime, Lazy<Double> serviceTime) {
        this.custID = custID;
        this.arrivalTime = arrivalTime;
        this.serviceTime = serviceTime.get(); 
    }
```

So that when I am populating my PQ, I will be able to:

<img width="766" alt="image" src="https://user-images.githubusercontent.com/77519191/161774350-d3e60a70-3ca1-4964-8ee8-74685c880db9.png">

I am not too sure if I am doing the right way. Thank you for your help!

**Comment by LauWyeRock** [comment]
yea thats what i have for my constructor too, just that for my this.serviceTime = serviceTime; (my servicetime attribute is a Lazy)

This is because if you call .get() in your constructor, all your customers will have a service time, but if you delay the .get() until after you determine whether the customer leaves or not, then only the customers that are gg to get served /waits will get a service time, which is what the question wants. If you look at the last test case, there are 10 customers but they only give 7 service times, because if your program runs properly 3 customers will leave, so they will not get a service time

**Comment by wzs02** [comment]
Hi @LauWyeRock, could I confirm whether you called the Lazy in the arrive and wait event, or whether you called in the serve event only? Thank you!

**Comment by hengyjj** [comment]
> yea thats what i have for my constructor too, just that for my this.serviceTime = serviceTime; (my servicetime attribute is a Lazy)
> 
> This is because if you call .get() in your constructor, all your customers will have a service time, but if you delay the .get() until after you determine whether the customer leaves or not, then only the customers that are gg to get served /waits will get a service time, which is what the question wants. If you look at the last test case, there are 10 customers but they only give 7 service times, because if your program runs properly 3 customers will leave, so they will not get a service time

Hi WyeRock. Thanks for your help! Just like to ask what did you take in for the first constructor that is meant for Simulate2 to Simulate4? It originally takes in a double service time, but as it is now switched to Lazy, I am not too sure how to write for this first constructor.

![image](https://user-images.githubusercontent.com/77519191/161872539-a92cb7ce-32be-4979-8c43-be2538544827.png)

**Comment by LauWyeRock** [comment]
@wzs02 i called it in the arrive and wait event yes

**Comment by LauWyeRock** [comment]
@hengyjj you can do a new Lazy<Double>(() -> 1.0)

**Comment by hengyjj** [comment]
> @hengyjj you can do a new Lazy(() -> 1.0)

Hi WyeRock, is this how you did it too? That we only call the serviceTime when it is an ArriveEvent and WaitEvent? I commented out my codes because there is some errors with my ArrivalEvent and WaitEvent constructors. Thank you for your help :)

![image](https://user-images.githubusercontent.com/77519191/161876249-6dc89314-6ba8-4d95-bbd9-1cd42dc01ab8.png)

**Comment by wzs02** [comment]
@LauWyeRock thanks for ur reply! Can I ask how u managed to call the lazy in the wait event and still have the correct assignment for waiting times? For level 5 test case 4, 4 waits before 5 but 5 is served first, so if you call in the lazy in the wait event, 4 will be assigned the wait time of 5 which is incorrect.

**Comment by hengyjj** [comment]
> @LauWyeRock thanks for ur reply! Can I ask how u managed to call the lazy in the wait event and still have the correct assignment for waiting times? For level 5 test case 4, 4 waits before 5 but 5 is served first, so if you call lazy in the wait event, 4 will be assigned the wait time of 5 which is incorrect.

I had the same issue right now as well..

**Comment by hengyjj** [comment]
> @LauWyeRock thanks for ur reply! Can I ask how u managed to call the lazy in the wait event and still have the correct assignment for waiting times? For level 5 test case 4, 4 waits before 5 but 5 is served first, so if you call lazy in the wait event, 4 will be assigned the wait time of 5 which is incorrect.

Hi wzs02, did you manage to resolve the issue?

**Comment by wzs02** [comment]
@hengyjj nope, I've tried posting and the response I got was to only call the lazy in the serve event, which mean I'll probably need to change the entire code :") Hopefully I can get a fix that allows me to call lazy in serve and wait like what @LauWyeRock mentioned.

**Comment by hengyjj** [comment]
> @hengyjj nope, I've tried posting and the response I got was to only call the lazy in the serve event, which mean I'll probably need to change the entire code :") Hopefully I can get a fix that allows me to call lazy in serve and wait like what @LauWyeRock mentioned.

Oh i just saw your post...

**Comment by hengyjj** [comment]
> @hengyjj nope, I've tried posting and the response I got was to only call the lazy in the serve event, which mean I'll probably need to change the entire code :") Hopefully I can get a fix that allows me to call lazy in serve and wait like what @LauWyeRock mentioned.

Do you mind replying here if you solve this issue successfully? Appreciate your help alot. Thank you.

**Comment by jamietan123** [comment]
I think you can just remove the part of your code in Wait, or else you may end up calling it twice for the same customer

**Comment by Brianliew77** [comment]
Yep for the first constructor I think u can change it to new Lazy(() -> 1.0) as mentioned above
